vrrpsyncd/vrrpsync.cpp

```diff
@@ -0,0 +1,640 @@
+#include <netlink/route/nexthop.h>
+#include <arpa/inet.h>
+#include <fstream>
+
+
+#include "logger.h"
+#include "dbconnector.h"
+#include "producerstatetable.h"
+#include "ipaddress.h"
+#include "netmsg.h"
+#include "linkcache.h"
+#include "exec.h"
+
+#include "vrrpsync.h"
+#include "warm_restart.h"
+#include "netmsg.h"
+
+using namespace std;
+using namespace swss;
+
+
+VrrpSync::VrrpSync(RedisPipeline *pipelineAppDB, DBConnector* cfgDb) :
+    m_vrrpTable(pipelineAppDB, APP_VRRP_TABLE_NAME),
+    m_cfgDeviceMetadataTable(cfgDb, CFG_DEVICE_METADATA_TABLE_NAME)
+{
+    m_nl_sock = nl_socket_alloc();
+    nl_connect(m_nl_sock, NETLINK_ROUTE);
+    rtnl_link_alloc_cache(m_nl_sock, AF_UNSPEC, &m_link_cache);
+    m_AppRestartAssist = new AppRestartAssist(pipelineAppDB, "vrrpsyncd", "vrrp", 2);	
+    if (m_AppRestartAssist)
+    {
+        m_AppRestartAssist->registerAppTable(APP_VRRP_TABLE_NAME, &m_vrrpTable);
+    }
+    if (getSystemMac(systemMac))
+    {
+        SWSS_LOG_NOTICE("Got system Mac %s", systemMac.c_str());
+    }
+
+    m_netLinkOnMsg = 0;
+    m_netLinkOnMsgUse = 0;
+    m_netLinkOnMsgLinkNew = 0;
+    m_netLinkOnMsgLinkNewUse = 0;
+    m_netLinkOnMsgLinkDel = 0;
+    m_netLinkOnMsgLinkDelUse = 0;
+    m_netLinkOnMsgAddrAdd = 0;
+    m_netLinkOnMsgAddrAddUse = 0;
+    m_netLinkOnMsgAddrDel = 0;
+    m_netLinkOnMsgAddrDelUse = 0;
+}
+
+VrrpSync::~VrrpSync()
+{
+    delete m_AppRestartAssist;
+}
+
+bool VrrpIsVrrpIntf(const char *ifname)
+{
+    if (!ifname)
+        return false;
+    
+    if (strncmp(ifname, "vrrp", 4))
+        return false;
+
+    return true;
+}
+
+string VrrpSync::VrrpIfindexToName(int ifindex)
+{
+    char ifname[MAX_ADDR_SIZE + 1] = {0};
+
+    if (rtnl_link_i2name(m_link_cache, ifindex, ifname, MAX_ADDR_SIZE) == NULL)
+    {
+        nl_cache_refill(m_nl_sock ,m_link_cache);
+        if (rtnl_link_i2name(m_link_cache, ifindex, ifname, MAX_ADDR_SIZE) == NULL)
+        {
+            return to_string(ifindex);
+        }
+
+    }
+
+    return string(ifname);
+}
+
+void VrrpSync::WriteToFile(const string& file, const string& value)
+{
+    FILE *pFile;
+
+    pFile = fopen(file.c_str(), "w");
+
+    if (pFile != NULL)
+    {
+        fputs(value.c_str(), pFile);
+        fclose(pFile);
+        SWSS_LOG_DEBUG("Updated the %s file with %s", file.c_str(), value.c_str());
+    }
+    else
+    {
+        SWSS_LOG_ERROR("Failed to open the %s file", file.c_str());
+    }
+
+}
+
+string VrrpSync::ReadLineFromFile(const string file, bool *exists)
+{
+    ifstream fp(file);
+    if (!fp.is_open())
+    {
+        SWSS_LOG_ERROR("File %s is not readable", file.c_str());
+        *exists = false;
+        return "";
+    }
+    string line="";
+    getline(fp, line);
+    if (line.empty())
+    {
+        SWSS_LOG_NOTICE("File %s is empty", file.c_str());
+    }
+    *exists = true;
+    fp.close();
+
+    return string(line);
+}
+
+
+void VrrpSync::VrrpUpdateNetdevFlags(string &ifname, int afi, bool up)
```

Comment:
Could you please explain why it is implemented in vrrpsyncd instead of macvlanmgrd?

Reply:
This is for vrrp parent interface. Is this same logic present in macvlanmgrd? Is so, it could be removed.